### ISSUE TITLE
elastic 5.1.1 master node always pinged time out, failed to form a cluster

### ISSUE THREAD

**Issue by sinsonglew** [elastic/elasticsearch#22189]
**Elasticsearch version**: 5.1.1

**Plugins installed**: [] only default

**JVM version**:  Java(TM) SE Runtime Environment (build 1.8.0_111-b14)

**OS version**: 2.6.32-504.12.2.02.el6.x86_64 #1 SMP Tue May 12 11:44:09 CST 2015 x86_64 x86_64 x86_64 GNU/Linux

**Description of the problem including expected versus actual behavior**:
if nodes greater than 3, the cluster is constantly add and remove nodes, leading to unavailable status.
From the following attached log, nodes ping each other timed out often
expected: a full-nodes started cluster

**Provide logs (if relevant)**:
```
[2016-12-15T19:39:17,648][INFO ][o.e.c.s.ClusterService   ] [node244] removed {{node237}{FmyPJDzMR2KRSl10jPGT0Q}{hrOcPGYBTOSEQU5J7hOwVA}{aa.bb.cc.dd}{aa.bb.cc.dd:9371},}, reason: master_failed ({node237}{FmyPJDzMR2KRSl10jPGT0Q}{hrOcPGYBTOSEQU5J7hOwVA}{aa.bb.cc.dd}{aa.bb.cc.dd:9371})
[2016-12-15T19:39:42,556][INFO ][o.e.d.z.ZenDiscovery     ] [node244] failed to send join request to master [{node237}{FmyPJDzMR2KRSl10jPGT0Q}{hrOcPGYBTOSEQU5J7hOwVA}{aa.bb.cc.dd}{aa.bb.cc.dd:9371}], reason [RemoteTransportException[[node237][aa.bb.cc.dd:9371][internal:discovery/zen/join]]; nested: IllegalStateException[failure when sending a validation request to node]; nested: NodeDisconnectedException[[node244][ee.ff.gg.hh:9371][internal:discovery/zen/join/validate] disconnected]; ]
[2016-12-15T19:40:02,565][INFO ][o.e.d.z.ZenDiscovery     ] [node244] failed to send join request to master [{node237}{FmyPJDzMR2KRSl10jPGT0Q}{hrOcPGYBTOSEQU5J7hOwVA}{aa.bb.cc.dd}{aa.bb.cc.dd:9371}], reason [RemoteTransportException[[node237][aa.bb.cc.dd:9371][internal:discovery/zen/join]]; nested: ConnectTransportException[[node244][ee.ff.gg.hh:9371] connect_timeout[30s]]; nested: IOException[Connection timed out: ee.ff.gg.hh/ee.ff.gg.hh:9371]; ]
[2016-12-15T19:40:25,433][INFO ][o.e.d.z.ZenDiscovery     ] [node244] failed to send join request to master [{node237}{FmyPJDzMR2KRSl10jPGT0Q}{hrOcPGYBTOSEQU5J7hOwVA}{aa.bb.cc.dd}{aa.bb.cc.dd:9371}], reason [RemoteTransportException[[node237][aa.bb.cc.dd:9371][internal:discovery/zen/join]]; nested: IllegalStateException[failure when sending a validation request to node]; nested: NodeDisconnectedException[[node244][ee.ff.gg.hh:9371][internal:discovery/zen/join/validate] disconnected]; ]
[2016-12-15T19:40:42,442][INFO ][o.e.d.z.ZenDiscovery     ] [node244] failed to send join request to master [{node237}{FmyPJDzMR2KRSl10jPGT0Q}{hrOcPGYBTOSEQU5J7hOwVA}{aa.bb.cc.dd}{aa.bb.cc.dd:9371}], reason [NodeDisconnectedException[[node237][aa.bb.cc.dd:9371][internal:discovery/zen/join] disconnected]]
[2016-12-15T19:41:03,531][INFO ][o.e.c.s.ClusterService   ] [node244] detected_master {node237}{FmyPJDzMR2KRSl10jPGT0Q}{hrOcPGYBTOSEQU5J7hOwVA}{aa.bb.cc.dd}{aa.bb.cc.dd:9371}, added {{node241}{gx3rLPdIQvSm_u6mLoIbtQ}{80MXRZOjSN6RAQ7gZGeUcQ}{xx.x.xx.x}{x.x.x.x:9371},{node237}{FmyPJDzMR2KRSl10jPGT0Q}{hrOcPGYBTOSEQU5J7hOwVA}{aa.bb.cc.dd}{aa.bb.cc.dd:9371},}, reason: zen-disco-receive(from master [master {node237}{FmyPJDzMR2KRSl10jPGT0Q}{hrOcPGYBTOSEQU5J7hOwVA}{aa.bb.cc.dd}{aa.bb.cc.dd:9371} committed version [115]])
[2016-12-15T19:41:09,541][INFO ][o.e.d.z.ZenDiscovery     ] [node244] master_left [{node237}{FmyPJDzMR2KRSl10jPGT0Q}{hrOcPGYBTOSEQU5J7hOwVA}{aa.bb.cc.dd}{aa.bb.cc.dd:9371}], reason [failed to ping, tried [3] times, each with  maximum [30s] timeout]
[2016-12-15T19:41:09,541][WARN ][o.e.d.z.ZenDiscovery     ] [node244] master left (reason = failed to ping, tried [3] times, each with  maximum [30s] timeout), current nodes: nodes: 
   {node243}{tEv4PrXSTdyecz2nClog4Q}{KQsJYnauSe2W0hxWPte4Iw}{xx.xx.xx.xx}{xx.xx.xx.xx:9371}
   {node239}{oWLoFJf-Rs-3q-kK1Y_2BQ}{dekHpx5zQASxKgIO21n1Yg}{xx.xx.xx.xx}{xx.xx.xx.xx:9371}
   {node244}{stop1h4jQc-zCvM2hBS56Q}{iV8QrIRiQ5y3cvW_en0sMA}{ee.ff.gg.hh}{ee.ff.gg.hh:9371}, local
   {node241}{gx3rLPdIQvSm_u6mLoIbtQ}{80MXRZOjSN6RAQ7gZGeUcQ}{xx.xx.xx.xx}{xx.xx.xx.xx:9371}
   {node242}{k5vCKHcjTSKTK-uhK_MtLw}{DMGU0D-8T-Si-ZpfqmiDTQ}{xx.xx.xx.xx}{xx.xx.xx.xx:9371}

[2016-12-15T19:41:09,542][INFO ][o.e.c.s.ClusterService   ] [node244] removed {{node237}{FmyPJDzMR2KRSl10jPGT0Q}{hrOcPGYBTOSEQU5J7hOwVA}{aa.bb.cc.dd}{aa.bb.cc.dd:9371},}, reason: master_failed ({node237}{FmyPJDzMR2KRSl10jPGT0Q}{hrOcPGYBTOSEQU5J7hOwVA}{aa.bb.cc.dd}{aa.bb.cc.dd:9371})
[2016-12-15T19:41:28,555][INFO ][o.e.c.s.ClusterService   ] [node244] master {new {node237}{FmyPJDzMR2KRSl10jPGT0Q}{hrOcPGYBTOSEQU5J7hOwVA}{aa.bb.cc.dd}{aa.bb.cc.dd:9371}}, removed {{node241}{gx3rLPdIQvSm_u6mLoIbtQ}{80MXRZOjSN6RAQ7gZGeUcQ}{xx.xx.xx.xx}{x.x.x.x:9371},}, added {{node237}{FmyPJDzMR2KRSl10jPGT0Q}{hrOcPGYBTOSEQU5J7hOwVA}{aa.bb.cc.dd}{aa.bb.cc.dd:9371},}, reason: zen-disco-receive(from master [master {node237}{FmyPJDzMR2KRSl10jPGT0Q}{hrOcPGYBTOSEQU5J7hOwVA}{aa.bb.cc.dd}{aa.bb.cc.dd:9371} committed version [120]])
[2016-12-15T19:41:34,564][INFO ][o.e.d.z.ZenDiscovery     ] [node244] master_left [{node237}{FmyPJDzMR2KRSl10jPGT0Q}{hrOcPGYBTOSEQU5J7hOwVA}{aa.bb.cc.dd}{aa.bb.cc.dd:9371}], reason [failed to ping, tried [3] times, each with  maximum [30s] timeout]
[2016-12-15T19:41:34,564][WARN ][o.e.d.z.ZenDiscovery     ] [node244] master left (reason = failed to ping, tried [3] times, each with  maximum [30s] timeout), current nodes: nodes: 
   {node243}{tEv4PrXSTdyecz2nClog4Q}{KQsJYnauSe2W0hxWPte4Iw}{xx.xx.xx.xx}{xx.xx.xx.xx:9371}
   {node239}{oWLoFJf-Rs-3q-kK1Y_2BQ}{dekHpx5zQASxKgIO21n1Yg}{xx.xx.xx.xx}{xx.xx.xx.xx:9371}
   {node244}{stop1h4jQc-zCvM2hBS56Q}{iV8QrIRiQ5y3cvW_en0sMA}{ee.ff.gg.hh}{ee.ff.gg.hh:9371}, local
   {node242}{k5vCKHcjTSKTK-uhK_MtLw}{DMGU0D-8T-Si-ZpfqmiDTQ}{xx.xx.xx.xx}{xx.xx.xx.xx:9371}

[2016-12-15T19:41:34,565][INFO ][o.e.c.s.ClusterService   ] [node244] removed {{node237}{FmyPJDzMR2KRSl10jPGT0Q}{hrOcPGYBTOSEQU5J7hOwVA}{aa.bb.cc.dd}{aa.bb.cc.dd:9371},}, reason: master_failed ({node237}{FmyPJDzMR2KRSl10jPGT0Q}{hrOcPGYBTOSEQU5J7hOwVA}{aa.bb.cc.dd}{aa.bb.cc.dd:9371})
[2016-12-15T19:41:55,575][INFO ][o.e.d.z.ZenDiscovery     ] [node244] failed to send join request to master [{node237}{FmyPJDzMR2KRSl10jPGT0Q}{hrOcPGYBTOSEQU5J7hOwVA}{aa.bb.cc.dd}{aa.bb.cc.dd:9371}], reason [RemoteTransportException[[node237][aa.bb.cc.dd:9371][internal:discovery/zen/join]]; nested: ConnectTransportException[[node244][ee.ff.gg.hh:9371] connect_timeout[30s]]; nested: IOException[Connection timed out: ee.ff.gg.hh/ee.ff.gg.hh:9371]; ]
[2016-12-15T19:42:15,586][INFO ][o.e.d.z.ZenDiscovery     ] [node244] failed to send join request to master [{node237}{FmyPJDzMR2KRSl10jPGT0Q}{hrOcPGYBTOSEQU5J7hOwVA}{aa.bb.cc.dd}{aa.bb.cc.dd:9371}], reason [RemoteTransportException[[node237][aa.bb.cc.dd:9371][internal:discovery/zen/join]]; nested: ConnectTransportException[[node244][ee.ff.gg.hh:9371] connect_timeout[30s]]; nested: IOException[Connection timed out: ee.ff.gg.hh/ee.ff.gg.hh:9371]; ]
```

**Comment by clintongormley** [comment]
Looks like you have networking problems. I don't know if you have any customs settings, or where you're running Elasticsearch etc.  More info would be helpful.

**Comment by sinsonglew** [comment]
@clintongormley , total 8 nodes, three of them are master eligible. we run elastic on virtual machines. It's definitely that network is fine. we tried elastic 1.x, it works very well.  elasticsearch.yml posted as follows:

# ======================== Elasticsearch Configuration =========================
#
# NOTE: Elasticsearch comes with reasonable defaults for most settings.
#       Before you set out to tweak and tune the configuration, make sure you
#       understand what are you trying to accomplish and the consequences.
#
# The primary way of configuring a node is via this file. This template lists
# the most important settings you may want to configure for a production cluster.
#
# Please see the documentation for further information on configuration options:
# <http://www.elastic.co/guide/en/elasticsearch/reference/current/setup-configuration.html>
#
# ---------------------------------- Cluster -----------------------------------
#
# Use a descriptive name for your cluster:
#
cluster.name: ${cluster.name}
#
# ------------------------------------ Node ------------------------------------
#
# Use a descriptive name for the node:
#
node.name: node237

#
# Add custom attributes to the node:
#
# node.rack: r1
#
# ----------------------------------- Paths ------------------------------------
#
# Path to directory where to store the data (separate multiple locations by comma):
#
#path.data: 
#
# Path to log files:
#
#path.logs:
#
# ----------------------------------- Memory -----------------------------------
#
# Lock the memory on startup:
#
bootstrap.memory_lock: true
bootstrap.seccomp: false
#
# Make sure that the `ES_HEAP_SIZE` environment variable is set to about half the memory
# available on the system and that the owner of the process is allowed to use this limit.
#
# Elasticsearch performs poorly when the system is swapping the memory.
#
# ---------------------------------- Network -----------------------------------
#
# Set the bind address to a specific IP (IPv4 or IPv6):
#
network.host: 10.xx.85.xx

transport.tcp.port: 9371

#
# Set a custom port for HTTP:
#
#http.port: 9200
#
# For more information, see the documentation at:
# <http://www.elastic.co/guide/en/elasticsearch/reference/current/modules-network.html>
#
# --------------------------------- Discovery ----------------------------------
#
# Pass an initial list of hosts to perform discovery when new node is started:
# The default list of hosts is ["127.0.0.1", "[::1]"]
#
discovery.zen.ping.unicast.hosts: ["10.xx.85.aa:9371", "10.xx.85.bb:9371", "10.xx.85.cc:9371"]
#
# Prevent the "split brain" by configuring the majority of nodes (total number of nodes / 2 + 1):
#
discovery.zen.minimum_master_nodes: 2

discovery.zen.ping_timeout: 17s
#
#discovery.zen.fd.ping_interval: ${discovery.zen.fd.ping_interval}
#
#discovery.zen.master_election.ignore_non_master_pings: ${discovery.zen.master_election.ignore_non_master_pings}
#
#discovery.zen.commit_timeout: ${discovery.zen.commit_timeout}
#
# For more information, see the documentation at:
# <http://www.elastic.co/guide/en/elasticsearch/reference/current/modules-discovery.html>
#
# ---------------------------------- Gateway -----------------------------------
#
# Block initial recovery after a full cluster restart until N nodes are started:
#
gateway.recover_after_nodes: 8
#
# For more information, see the documentation at:
# <http://www.elastic.co/guide/en/elasticsearch/reference/current/modules-gateway.html>
#
# ---------------------------------- Various -----------------------------------
#
# Disable starting multiple nodes on a single system:
#
# node.max_local_storage_nodes: 1
#
# Require explicit names when deleting indices:
#
# action.destructive_requires_name: true

**Comment by clintongormley** [comment]
1.x used multicast, you may well find that it was using different network cards etc than the ones configured.  It still looks to me like you have a networking issue here, or possibly different config on one of the nodes?

**Comment by georgecao** [comment]
@clintongormley , I work with @sinsonglew on the same team.  Does ElasticSearch 5.1.1 require specific network cards? What's your concerns and how we check it out?  If we have 10 nodes to form one cluster, what's elasticsearch.yml should look like?  It's wired that we reduce the number of nodes from 10 to 3, then the cluster becomes stable, but increase the number of nodes from 3 to 4 or 5, it happens again. 

Any idea?
Thanks in advance.

**Comment by bleskes** [comment]
@georgecao I think there are a couple of things going on.

First you say you have 10 nodes, I presume you don't run dedicated master nodes. If that's the case then setting `discovery.zen.minimum_master_nodes` to `2` is dangerous as your 10 nodes can split to form 3 clusters of 3 nodes each. You should set it to `6` (a majority). That said, it's not clear to me why this explains network disconnects.

It seems the nodes discover the master fine (and try to join it), so I don't think it's a question of multicast based discovery. 

It seems something is off with the network there are a lot of timeouts and disconnect errors in the log you shared. Since you masked the IPs (good!) it's hard for me to check they all make sense. In ES each node has the local ip it binds to and the IP it publishes (i.e., the IP people should use to connect to it). Can you check those make sense?

Finally, I was wondering how you ended up with 17s in `discovery.zen.ping_timeout: 17s` ?

**Comment by sinsonglew** [comment]
@Boaz, at first, 10 nodes without dedicated roles settings were experimented, the minimum master nodes set up with 6, we can't get a full started cluster, so we have to decrease the total nodes to 3, it works, then 4, it failed. We have another dc, so 8 nodes with 3 master eligible got tried again, it repeated. We can't figure out where there is something wrong. Thanks!

发自我的 Windows Phone
________________________________

**Comment by bleskes** [comment]
@sinsonglew everything points at network issues. Can you check the IPs of the published address?

**Comment by sinsonglew** [comment]
@bleskes, @clintongormley, we downgrade netty4 to netty3, problem sovled, it's not network issue, please take a concern.

**Comment by bleskes** [comment]
> we downgrade netty from 4 to 3 by setting transport.type=netty3
http.type=netty3 , problem sovled, it's not a network issue.

Thank you for this clue. We will investigate and get back to you as soon as possible.

**Comment by jasontedor** [comment]
Can you share some more details:
 - node specs (especially heap size but general specs as well)
 - from your kernel version I think that you're running RHEL 6, can you confirm?
 - logs from a node and the master (there is no need to mask the IP addresses, from your configuration I see that you're in 10.0.0.0/8 which is a private IP address space; masking will just make it harder-than-necessary to read)
 - are the nodes accessible over HTTP (e.g., `GET /`)?
 - what is going on on the cluster when this happens (heavy indexing, searching, etc.)?

**Comment by georgecao** [comment]
> node specs (especially heap size but general specs as well)

In **jvm.options**, we specify the heap size and enable gc logs and log to a file by setting
```
-Xms25g
-Xmx25g
# ensure the directory exists
-Xloggc:/path/to/gc/log.file
```
and keep the other configurations as same as the default ones.

> from your kernel version I think that you're running RHEL 6, can you confirm?

`lsb_release -a` prints **CentOS release 6.4 (Final)**

> are the nodes accessible over HTTP (e.g., GET /)?

Yes, but very slow.

> what is going on on the cluster when this happens (heavy indexing, searching, etc.)?

The cluster do nothing special. This happens just after we install the ElasticSearch 5.1.1, then we try to start the cluster, which means there are no indexing, no searching and no data on any node.

> logs from a node and the master (there is no need to mask the IP addresses, from your configuration I see that you're in 10.0.0.0/8 which is a private IP address space; masking will just make it harder-than-necessary to read)

Indeed, the private IP is safe to publish, but the hostname is not.   Will update logs including the private IP.

**Comment by jasontedor** [comment]
> Yes, but very slow.

This is very suspicious, and I suspect that if we can understand this we might be able to unlock what is going on here.

How many cores do your machines have?

Do you see anything in the GC logs?

Can you start the JVMs with these flags?

```
-XX:+PrintGCDetails
-XX:+PrintGCTimeStamps
-XX:+PrintGCDateStamps
-XX:+PrintClassHistogram
-XX:+PrintTenuringDistribution
-XX:+PrintGCApplicationStoppedTime
-XX:+PrintGCApplicationConcurrentTime
-Xloggc:/path/to/gc/log.file
```

**Comment by aqlu** [comment]
We upgrade 2.3.2 to 5.1.1, And encountered the same problem with @sinsonglew

**Comment by aqlu** [comment]
@sinsonglew we tried to downgrade netty from 4 to 3, but can not resolved this problem.

We found some phenomenon:
1. when the throughput is relatively low, there is no node left. 
2. Master node is never left.

I tried to setting `discovery.zen.fd.ping_retries: 6`, but has no effect.

**Elasticsearch version:** 5.1.1
**Plugins installed:** [x-pack]  disabled security plugin
**JVM version:** Java(TM) SE Runtime Environment (build 1.8.0_74-b02)
**OS version:** CentOS release 6.7 (Final) 2.6.32-573.8.1.el6.x86_64 #1 SMP Tue Nov 10 18:01:38 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux

there is my config:
```yml
cluster.name: prod_elasticsearch
node.name: es-log-193
path.data: /home/elasticsearch/data
network.host: 0.0.0.0
discovery.zen.ping.unicast.hosts: ["192.168.65.188","192.168.65.189","192.168.65.190","192.168.65.191","192.168.65.187","192.168.65.192"]
xpack.monitoring.collection.interval: 30s
xpack.monitoring.history.duration: 3d
http.cors.enabled: true
http.cors.allow-origin: '*'
xpack.security.enabled: false
thread_pool.index.queue_size: 8000
thread_pool.search.queue_size: 8000
thread_pool.bulk.queue_size: 8000
path.repo: ["/home/elasticsearch/ES_backup"]
discovery.zen.fd.ping_retries: 6
```

**Comment by RobertTheProfessional** [comment]
@sinsonglew @aqlu Just upgraded our 10 node cluster from 2.4.1 and suffering what seems to be similar issues.

I've downgraded the netty clients on both the transports and http clients to version 3 from 4 and adjusted the keepalive timers on the all of the hosts but my nodes are still suffering the pinging issues every few hours.

```
[2016-12-27T19:25:07,916][INFO ][o.e.d.z.ZenDiscovery     ] [Data Six] master_left [{Master One}{VLblVsoKQnK82vCcn8aU7w}{pIwD1j5ESESvFQ0ZCoe8GQ}{x.x.x.7}{x.x.x.7:9300}{rack=Masters}], reason [failed to ping, tried [3] times, each with  maximum [30s] timeout]
[2016-12-27T19:25:07,917][WARN ][o.e.d.z.ZenDiscovery     ] [Data Six] master left (reason = failed to ping, tried [3] times, each with  maximum [30s] timeout), current nodes: nodes: 
	{Data Four}{v4GIb5n7SSess4fUCBCbsQ}{PirkGa3rQ1KtX89FNC3xuQ}{x.x.x.11}{x.x.x.11:9300}{rack=Search Load Balancers}
	{Data Eight}{4_ly1NagS7-2fp9z5KiIHQ}{AKs98d44TSCS0BOYhA4Idg}{x.x.x.22}{x.x.x.22:9300}{rack=Data Stores}
	{Data Five}{sxEkv3nBSjaa3OkNN6kZnQ}{Q6ReG88NSQi5VButwrIDCA}{x.x.x.12}{x.x.x.12:9300}{rack=Data Stores}
	{Kibana}{Raicx_zzTyGqedAwRCsxNw}{obZmcC9BQA2dFtXQXpx4ug}{x.x.x.15}{x.x.x.15:9300}{rack=Visualization Server}
	{Data Nine}{9C9GxU7LQgyixJtvWcpc-Q}{WXtLtBasTJ2kFCnZNYlQqw}{x.x.x.23}{x.x.x.23:9300}{rack=Data Stores}
	{Data Seven}{52kTpnu9SeWTI_smXBu3tA}{_q9bPoezSrOYGVS8nQoXLQ}{x.x.x.21}{x.x.x.21:9300}{rack=Data Stores}
	{Logstash One}{Pg0J5PhEQU-6dRaoV8tXxQ}{PcY7w0fARiW2p43VRvC6YA}{x.x.x.17}{x.x.x.17:9300}{rack=Ingestion Server}
	{Data Ten}{tJUQk8MiQqC4QqJI-hWmmA}{d5d5YkVjRgqM7OieRxtedQ}{x.x.x.24}{x.x.x.24:9300}{rack=Data Stores}
	{Master Two}{j4B-jvfeT2CAESG5sElJXg}{1lVXxxDXTde5VkUF0UH29Q}{x.x.x.8}{x.x.x.8:9300}{rack=Masters}
	{Master Three}{iQzn2UDZQ-27zZIPrIW-8Q}{nSmaCPW5RWSy7L9L5EaYow}{x.x.x.9}{x.x.x.9:9300}{rack=Masters}
	{Data Six}{6iptJ5_wQ4aHDs1DYwc2SQ}{Z7IJ_XrmTHGK-FWJEmUgsw}{x.x.x.13}{x.x.x.13:9300}{rack=Data Stores}, local
	{Data One}{W9KXL56rRJeXDsyBV-J1nA}{uz8gReByRmmZsvz52saFkQ}{x.x.x.4}{x.x.x.4:9300}{rack=Data Stores}
	{Data Two}{cba4LbEiTDSzWGEvlvLwwA}{KGkQe5I1RSG5Q3bVH7Mrrw}{x.x.x.5}{x.x.x.5:9300}{rack=Data Stores}
	{Data Three}{wCjYdTB_RTO2QkqmCoE40w}{gJHQonRMSXSjiD43PcVdXw}{x.x.x.6}{x.x.x.6:9300}{rack=Data Stores}
`````

**Comment by sinsonglew** [comment]
@aqlu, what's your setting for **discovery.zen.minimum_master_nodes**, you said when throughput is low, it works fine, vice versa. it is how high the throughput is that make cluster out of function. Did  cpu, memory usage are in safety line, how complex your query are ? Try optimize your query may be the first option.

**Comment by sinsonglew** [comment]
@RobertTheProfessional , you said every few hours the cluster will lose nodes. When that happened, can you provide any other info about business query frequency, lcpu load and so on.

Also try a different master node to exclude machine issue.

**Comment by RobertTheProfessional** [comment]
@sinsonglew The nodes are on an OpenStack virtualization platform. We have three master nodes at 4 vCPUs, 16GB RAM and 4GB heap locked to the ES process. There's ten data nodes are 16 vCPUs, 64GB RAM with 31GB of RAM locked to the ES process. Currently we're indexing at a rate of 5,000 documents a second and as a result the heap on a few of the nodes tend to average around 27GB before the garbage collector kicks in. Previously we had a six node cluster with a three master/three data node but the cluster was failing constantly due to a high index rate. The index that's been causing the issues is averaging 38 million documents that are loaded dynamically through Logstash's JSON filter.

I attached a screenshot of one of the problematic nodes, although most of the nodes suffer from a failure to ping the master.

<img width="1756" alt="screen shot 2016-12-27 at 8 47 02 pm" src="https://cloud.githubusercontent.com/assets/1852571/21511986/b126e112-cc75-11e6-92eb-964b8117d60b.png">

**Comment by aqlu** [comment]
@sinsonglew here is my nodes info, 5 master and data node,  2 data only node:
```
192.168.65.193 59 89 16 8.64 8.29 8.17 mdi - es-log-193
192.168.65.189 32 92 20 5.77 6.23 5.90 mdi - es-log-189
192.168.65.190 72 91 20 4.94 5.48 5.31 mdi - es-log-190
192.168.65.188 31 90 30 6.20 6.32 5.87 di  - es-log-188
192.168.65.191 42 90 29 5.26 5.66 5.34 di  - es-log-191
192.168.65.187 51 91 20 4.59 5.22 5.18 mdi * es-log-187
192.168.65.192 15 91 18 6.06 6.33 5.88 mdi - es-log-192
```
We setting **discovery.zen.minimum_master_nodes ** is **3** . This ES cluster is our log cluster, usually use bulk index, query is less. And this cluster working fine at ES v2.3.2.
CPU and memory are in noraml range:
![image](https://cloud.githubusercontent.com/assets/5010463/21512300/ba95ecd6-cce6-11e6-824c-45506c503821.png)

@RobertTheProfessional It seems more guys encounter with this trouble.

@clintongormley @bleskes @jasontedor 
Do you have any ideas?

**Comment by RobertTheProfessional** [comment]
@aqlu Yeah this issue cropped up immediately after upgrading to 5.1.1. Judging by both of our statistics, I'm starting to believe the garbage collection process is getting in the way of the node pinging the master node.

<img width="1697" alt="screen shot 2016-12-27 at 9 47 04 pm" src="https://cloud.githubusercontent.com/assets/1852571/21512832/3e9cc224-cc7f-11e6-96c8-9f1e029b33ac.png">

**Comment by sinsonglew** [comment]
@aqlu , how did you assign shards allocation ? what's your setting of  index **refresh_interval** ? 
try to adjust **refresh_interval** bigger.

**Comment by aqlu** [comment]
@sinsonglew 
here is my refresh and shards setting:
```
"refresh_interval": "5s"
"number_of_shards": "14"
"number_of_replicas": "1"
```
I can try to set up value of **refresh_interval**， But I think this is not root cause of the problem.

**Comment by dropje86** [comment]
Same problem here. Have a test cluster running with both 2.4.3 and 5.1.1. Both are benchmarked with a very simple Go script to ingest as much data as possible. We can get 2.4.3 on it's knees without any issues regarding node disconnections, just high load. 5.1.1 intermittently has the same issues as described multiple times before in this thread.

Setup with both versions the same:

3 - master nodes
3 - client nodes
16 - data nodes

```
discovery.zen.fd.ping_interval: 1s
discovery.zen.fd.ping_timeout: 5s
discovery.zen.fd.ping_retries: 3
discovery.zen.minimum_master_nodes: 2
```

**Comment by jasontedor** [comment]
@dropje86 Are you running any monitoring agent that periodically makes stats calls to Elasticsearch (e.g., to `/_nodes/stats`)?

**Comment by dropje86** [comment]
@jasontedor yes, in the case of 2.4.3 we run marvel-agent and with 5.1.1 we run with x-pack installed.

**Comment by jasontedor** [comment]
@dropje86 Sadly, I suspect that you're being bitten by #22285; the fix is #22317 (and relates to #22314) and will be released in 5.1.2.

**Comment by dropje86** [comment]
I would be fine with that ;) Came across that issue before I saw this one, but I don't see any related exceptions in the master logs. I'm just getting a ping timeout to the master from a suffering data node and the active master reports transport disconnections to that data node.

**Comment by RobertTheProfessional** [comment]
@jasontedor Interestingly, I'm using a modified version of Kopf that makes constant `/_nodes/stats` calls to all of the nodes in my cluster. That could be the reason as to why the nodes are having transport issues with the master nodes?

**Comment by jasontedor** [comment]
I think that this is addressed by #22452; please reopen if not.

**Comment by jerome-918** [comment]
Give an advision: please make sure the node opens the port 9300.
